File: multilanguage/demo/demo_app.cpp
#include <iostream>
#include<uuid/uuid.h>   // For UUID creation.

using namespace std;

void come_back() {
  cout << "Back in CPP. !!!" << endl;
}


class Sapphire {

  uuid_t uid;
  char* app_string = (char*)"Hello DCAP from CPP. !!!"; 
  
  public:
    
    char* get_app_string()
    {
      return app_string;
    }

    void  set_sapphire_instance_uuid(char* id)
    {
       uuid_parse(id, uid);
    }
    char* get_sapphire_instance_uuid()
    {
      char* id = new char[100];
      uuid_unparse(uid,id);
      return id;
    }
    
    
    char* sapphire_createInstance();
    void  sapphire_deleteInstance(char*);
    
    void sapphire_helloWorld(char*);
    void sapphire_getObjCount();
    void sapphire_incObjCount();
};

char* Sapphire::sapphire_createInstance()
{
  char* uuid = NULL;
  cout << "Creating the Sapphire instance from CPP. !!!" << endl;
  return uuid;
}

void Sapphire::sapphire_deleteInstance(char* id)
{
  cout << "Deleting the Sapphire instance from CPP. !!! UUID: " << id << endl;
}

void Sapphire::sapphire_helloWorld(char* app_string)
{
  cout << "Hello World from CPP. !!!" << endl;
}

void Sapphire::sapphire_getObjCount()
{
  cout << "Inside CPP getObjCount() !!!" << endl;
}

void Sapphire::sapphire_incObjCount()
{
  cout << "Inside CPP sapphire_incObjCount() !!!";
}

int main() {

  cout << "Inside CPP main function. !!!" << endl;

  Sapphire *obj = new Sapphire();
  obj->sapphire_helloWorld(obj->get_app_string());
  delete obj;
  
  come_back();
}

File: multilanguage/demo/demo_app.go
package main

import "fmt"

func come_back() {
  fmt.Println("Back in Golang. !!!")
}

type DCAP_Sapphire interface {
  sapphire_helloWorld(string)
  sapphire_getObjCount()
  sapphire_incObjCount()
}

type Sapphire struct {
}

func (f Sapphire) sapphire_helloWorld(string) {
  fmt.Println("Inside golang helloWorld() !!!")
}

func (f Sapphire) sapphire_getObjCount() {
  fmt.Println("Inside golang getObjCount() !!!")
}

func (f Sapphire) sapphire_incObjCount() {
  fmt.Println("Inside golang incObjCount() !!!")
}

func newSapphire() DCAP_Sapphire {
  return &Sapphire{}
}

func main() {
 fmt.Println("Inside Golang main function. !!!")
 
 // New Inerface object
 obj := newSapphire()

 app_string := "Hello DCAP from Golang. !!!"
 obj.sapphire_helloWorld(app_string)
 
 obj.sapphire_getObjCount()
 obj.sapphire_incObjCount() 
 
 come_back()
}   

